File: common/src/main/java/com/ssblur/scriptor/helpers/targetable/InventoryTargetable.java
package com.ssblur.scriptor.helpers.targetable;

import net.minecraft.world.Container;
import net.minecraft.world.item.ItemStack;
import org.jetbrains.annotations.Nullable;

import java.util.function.Predicate;

public interface InventoryTargetable {
  @Nullable
  Container getContainer();

  int getTargetedSlot();
  void setTargetedSlot(int slot);

  default void useFirstMatchingSlot(Predicate<ItemStack> predicate) {
    int slot = getFirstMatchingSlot(predicate);
    if(slot >= 0)
      setTargetedSlot(slot);
  }

  default int getFirstMatchingSlot(Predicate<ItemStack> predicate) {
    if(getContainer() != null)
      for(int i = 0; i < getContainer().getContainerSize(); i++)
        if(predicate.test(getContainer().getItem(i))) {
          return i;
        }
    return -1;
  }
  default void useFirstFilledSlot() {
    int slot = getFirstFilledSlot();
    if(slot >= 0)
      setTargetedSlot(slot);
  }

  default int getFirstFilledSlot() {
    if(getContainer() != null)
      for(int i = 0; i < getContainer().getContainerSize(); i++) {
        if (!getContainer().getItem(i).isEmpty()) {
          setTargetedSlot(i);
        }
      }
    return -1;
  }
}

File: common/src/main/kotlin/com/ssblur/scriptor/helpers/targetable/InventoryTargetable.kt
package com.ssblur.scriptor.helpers.targetable

import net.minecraft.world.Container
import net.minecraft.world.item.ItemStack
import java.util.function.Predicate

interface InventoryTargetable {
  val container: Container?

  var targetedSlot: Int

  fun shouldIgnoreTargetedSlot(): Boolean {
    return targetedSlot == -1 && container != null
  }

  fun useFirstMatchingSlot(predicate: Predicate<ItemStack>) {
    val slot = getFirstMatchingSlot(predicate)
    if (slot >= 0) targetedSlot = slot
  }

  fun getFirstMatchingSlot(predicate: Predicate<ItemStack>): Int {
    if (container != null) for (i in 0 until container!!.containerSize) if (predicate.test(container!!.getItem(i))) {
      return i
    }
    return -1
  }

  val firstFilledSlot: Int
    get() = getFirstMatchingSlot { item: ItemStack -> !item.isEmpty }

  fun getFirstMatchingSlot(itemStack: ItemStack): Int {
    return getFirstMatchingSlot { item: ItemStack ->
      (item.isEmpty
          || (ItemStack.isSameItemSameComponents(item, itemStack)
          && (item.count + itemStack.count) <= item.maxStackSize
          ))
    }
  }

  fun getFirstMatchingSlotNotEmpty(itemStack: ItemStack): Int {
    return getFirstMatchingSlot { item: ItemStack ->
      (ItemStack.isSameItemSameComponents(item, itemStack)
          && (item.count + itemStack.count) <= item.maxStackSize)
    }
  }

  fun useFirstFilledSlot() {
    val slot = firstFilledSlot
    if (slot >= 0) targetedSlot = slot
  }
}
